apisix/plugins/traffic-split.lua

```diff
@@ -0,0 +1,294 @@
+    -- the upstream of `route` is used by default.
+    default = {
+        {
+            weighted_upstream = 1
+        }
+    },
+    minItems = 1,
+    maxItems = 20
+}
+
+
+local schema = {
+    type = "object",
+    properties = {
+        rules = {
+            type = "array",
+            items = {
+                type = "object",
+                properties = {
+                    match = match_schema,
+                    upstreams = upstreams_schema
+                }
+            }
+        }
+    }
+}
+
+local plugin_name = "traffic-split"
+
+local _M = {
+    version = 0.1,
+    priority = 966,
+    name = plugin_name,
+    schema = schema
+}
+
+function _M.check_schema(conf)
+    local ok, err = core.schema.check(schema, conf)
+
+    if not ok then
+        return false, err
+    end
+
+    return true
+end
+
+
+local function parse_domain_for_node(node)
+    if not ipmatcher.parse_ipv4(node) and not ipmatcher.parse_ipv6(node) then
+        local ip, err = init.parse_domain(node)
+        if ip then
+            return ip
+        end
+
+        if err then
+            return nil, err
+        end
+    end
+
+    return node
+end
+
+
+local function set_upstream(upstream_info, ctx)
+    local nodes = upstream_info.nodes
+    local new_nodes = {}
+    for addr, weight in pairs(nodes) do
+        local node = {}
+        local ip, port, host
+        host, port = core.utils.parse_addr(addr)
+        ip = parse_domain_for_node(host)
+        node.host = ip
+        node.port = port
+        node.weight = weight
+        table_insert(new_nodes, node)
+
+        -- Currently only supports a single upstream of the domain name.
+        -- When the upstream is `IP`, do not do any `pass_host` operation.
+        if not core.utils.parse_ipv4(host) and not core.utils.parse_ipv6(host) then
+            local pass_host = upstream_info.pass_host or "pass"
+            if pass_host == "pass" then
+                ctx.var.upstream_host = ctx.var.host
+                break
+            end
+
+            if pass_host == "rewrite" then
+                ctx.var.upstream_host = upstream_info.upstream_host
+                break
+            end
+
+            ctx.var.upstream_host = host
+            break
+        end
+    end
+    core.log.info("upstream_host: ", ctx.var.upstream_host)
+
+    local up_conf = {
+        name = upstream_info.name,
+        type = upstream_info.type,
+        nodes = new_nodes,
+        timeout = {
+            send = upstream_info.timeout and upstream_info.timeout.send or 15,
+            read = upstream_info.timeout and upstream_info.timeout.read or 15,
+            connect = upstream_info.timeout and upstream_info.timeout.connect or 15
+        }
+    }
+
+    local ok, err = upstream.check_schema(up_conf)
+    if not ok then
+        return 500, err
+    end
+
+    local matched_route = ctx.matched_route
+    upstream.set(ctx, up_conf.type .. "#route_" .. matched_route.value.id,
```

Comment:
`up_conf.type .. "#route_" .. matched_route.value.id` would produce something like "roundrobin#route_1", which will further to be used as the upstream key in https://github.com/apache/apisix/blob/1f72f6726055e81cc245409d498084740cb3e1af/apisix/balancer.lua#L242.
Will it be a problem? In my opinion, it will cause different upstreams to use the same key when picking node.

Reply:
Yes, you are right. We should provide an unique key for the generated upstream.

Reply:
Yes, thank you very much for your comments.

Reply:
@spacewander 
I tried it, but couldn't find a good way to generate unique keys, because in the plug-in, except for the nodes information of upstream, the other configuration information is the same.
Do you have any good suggestions for this problem?

Reply:
What about generating a UUID as a suffix? We don't need to reuse in this occasion.

Reply:
I have tried the uuid method, but this will result in a situation where all requests go to only one upstream.

```
got: '1982, 1982, 1982, 1982, 1982
# '
#     expected: '1980, 1981, 1981, 1982, 1982
```

Reply:
Look like it is because the wrr is used as the picker? If we use random, this isn't a problem. Since the weight and node are changed per request, it is not safe to reuse the picker.

Reply:
I think this does not affect the results of our choice of different upstream. Here wrr is used as the picker the same as the picker in upstream nodes. They are able to work normally.

Reply:
I think it again. Look like we can assign a virtual id to each upstream of the `upstreams` field, then use the vid as the suffix, so that we can reuse the upstream without messing up.